File: examples/BrisaLight/switchPower.cpp
#include "switchPower.h"

using namespace brisa;
using namespace brisa::upnp;
using namespace brisa::upnp::device;

SwitchPower::SwitchPower() :
        Service(SERVICE_TYPE,
                     SERVICE_ID,
                     SERVICE_XML_PATH,
                     SERVICE_CONTROL,
                     SERVICE_EVENT_SUB)
{
}

OutArgument* SwitchPower::getstatus(InArgument *const inArguments, Action *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    OutArgument *outArgs = new OutArgument();
    outArgs->insert("ResultStatus", this->getVariable("Status")->
                    getAttribute(StateVariable::Value));
    return outArgs;
}

OutArgument* SwitchPower::gettarget(InArgument *const inArguments, Action *const action)
{
    Q_UNUSED(inArguments)
    Q_UNUSED(action)
    OutArgument *outArgs = new OutArgument();
    outArgs->insert("RetTargetValue", this->getVariable("Target")->
                    getAttribute(StateVariable::Value));
    return outArgs;
}

OutArgument* SwitchPower::settarget(InArgument *const inArguments, Action *const action)
{
    Q_UNUSED(action)
    this->getVariable("Target")->setAttribute(StateVariable::Value,
                                                   (*inArguments)["NewTargetValue"]);
    this->getVariable("Status")->setAttribute(StateVariable::Value,
                                                   (*inArguments)["NewTargetValue"]);
    return new OutArgument();
}

File: platforms/java/jbrisa-examples/JBrisaBinaryLight/src/binarylight/SwitchPower.java
package binarylight;

import java.util.ArrayList;

import upnp.device.core.BrisaService;
import upnp.device.model.BrisaArgument;

public class SwitchPower extends BrisaService {
	
	private final static String SERVICE_ID = "SwitchPower";
	private static final String SERVICE_TYPE = "urn:schemas-upnp-org:service:SwitchPower:1";
	//private static final String SERVICE_XML_PATH = "/SwitchPower-scpd.xml";
	//private static final String SERVICE_EVENT_SUB = "/SwitchPower/eventSub";
	//private static final String SERVICE_CONTROL = "/SwitchPower/control";

	public SwitchPower() {
		super(SERVICE_ID, SERVICE_TYPE);
	}
	
	/**
	 * Retorna o Status da Lampada Binaria
	 * 
	 * @param in Argumentos de entrada
	 * @return Argumentos de saida
	 */
	public ArrayList<BrisaArgument> getStatus(ArrayList<BrisaArgument> inArgs) {
		
		System.out.println("getStatus");
		
		BrisaArgument out = new BrisaArgument("ResultStatus", "out", "Status");	
		out.setValue(this.getStateVariable(out.getRelatedStateVariable()).getValue());
		
		ArrayList<BrisaArgument> outArgs = new ArrayList<BrisaArgument>();
		outArgs.add(out);
		
		System.out.println("in Args: " + inArgs);
		System.out.println("out Args: " + outArgs);
		
	    return outArgs;
	}

	/**
	 * Retorna o Status da Lampada Binaria
	 * 
	 * @param in Argumentos de entrada
	 * @return Argumentos de saida
	 */
	public ArrayList<BrisaArgument> getTarget(ArrayList<BrisaArgument> inArgs)
	{
		System.out.println("getTarget");
	    		
		BrisaArgument out = new BrisaArgument("RetTargetValue", "out", "Target");
		out.setValue(this.getStateVariable(out.getRelatedStateVariable()).getValue());

		ArrayList<BrisaArgument> outArgs = new ArrayList<BrisaArgument>();
		outArgs.add(out);
		
		System.out.println("in Args: " + inArgs);
		System.out.println("out Args: " + outArgs);
		
	    return outArgs;
	}

	/**
	 * Muda o estado da lampada binaria
	 * 
	 * @param in Argumentos de entrada
	 * @return Argumentos de saida (vazio)
	 */
	public ArrayList<BrisaArgument> setTarget(ArrayList<BrisaArgument> inArgs)
	{
		System.out.println("setTarget");
		
		BrisaArgument in = inArgs.get(0);
		this.getStateVariable("Target").setValue(in.getValue());
		this.getStateVariable("Status").setValue(in.getValue());
		
		System.out.println("in Args: " + inArgs);
		System.out.println("out Args: <vazio>");
		
	    return new ArrayList<BrisaArgument>(); //Sem out Arguments
	}
	
	

}
